File: 092. Reverse Linked List II/092. Reverse Linked List II.cpp
//主要思路是记录下翻转的链表的头和尾，分别赋值。
//方法来源：https://blog.csdn.net/a2331046/article/details/51326259


ListNode* reverseBetween(ListNode* head, int m, int n) {
    ListNode* pre = NULL;
    ListNode* cur = head;
    ListNode* tmpTail;
    for (int i = 1 ; i < m;i++){
        pre = cur;
        cur = cur->next;
    }
    tmpTail = cur;
    for (int i = m ; i <= n ; i++){
        ListNode* next = cur->next;
        cur->next = pre;
        pre = cur;
        cur = next;
    }
    if (m==1) head = pre;
    else tmpTail->next->next = pre;
    
    tmpTail->next = cur;
    return head;
}

File: 092. Reverse Linked List II/092. Reverse Linked List II.py
#进行一遍遍历，把第m到n个元素进行翻转，即依次插入到第m个节点的头部。
#方法来源：https://blog.csdn.net/fuxuemingzhu/article/details/80794665 
class Solution(object):
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        count = 1
        root = ListNode(0)
        root.next = head
        pre = root
        while pre.next and count < m:
            pre = pre.next
            count += 1
        if count < m:
            return head
        mNode = pre.next
        curr = mNode.next
        while curr and count < n:
            next = curr.next
            curr.next = pre.next
            pre.next = curr
            mNode.next = next
            curr = next
            count += 1
        return root.next
